### PR TITLE
WI #2605 Fix wrong URI management

### LINKED ISSUE #2605: Incorrect URI management in notifications PublishDiagnostics and MissingCopies
**Describe the bug**
When sending notifications `PublishDiagnostics `and `MissingCopies `to the client, the server does not provide escaped URI (for instance, blanks are not replaced by `%20`).
The client is consequently not able to retrieve the related document when receiving the notification.

**To Reproduce**
Open a document with Cobol syntax errors from a Repository Workspace containing some blanks: no diagnostics are shown.

**Expected behavior**
The diagnostics should be visible.

**Technical**
When creating the URI parameter in the notification, the server uses `ToString `on the original URI which may result in a wrong value (blanks are not escaped by `ToString`). 

**How to test automatically**
Not relevant

See also #2606

### PR DESCRIPTION
**Pull request by efr15**
Fixes #2605

Side effects: LSR tests must be adapted accordingly (URI 'file:///...' changed to 'file:/...')

### CODE REVIEW

File: TypeCobol.LanguageServer/TypeCobolServer.cs [hunk 1]
```diff
@@ -239,6 +239,18 @@ protected DocumentContext GetDocumentContextFromStringUri(string uri, Workspace.
...
+        /// Gets the original URI (which was set by the client)
+        /// DON'T use ToString() as it returns the canonically unescaped form of the URI
+        /// (it may cause issue if the path contains some blanks which need to be escaped)
+        /// </summary>
+        /// <param name="uri"></param>
+        /// <returns></returns>
+        private static string GetOriginalUri(Uri uri)
+        {
+            return uri.OriginalString;
+        }
```
Comment (fm-117) [inline TypeCobol.LanguageServer/TypeCobolServer.cs:252]: Please move that to https://github.com/TypeCobolTeam/TypeCobol/blob/develop/TypeCobol.LanguageServer/Utilities/Extensions.cs and make it extension an extension method.

Comment (fm-117) [reply TypeCobol.LanguageServer/TypeCobolServer.cs:252]: I think it's better to create a new constructor for `TextDocumentIdentifier` accepting an `Uri` parameter and call `OriginalString` in it. Duplicate the call to `OriginalString` for the `diagParameter.uri` assignment and remove this method.
You can keep the comment about using the original value sent by the client in the new constructor.

**Review by fm-117 — CHANGES_REQUESTED** [review]

**Comment by efr15** [comment]
Adding a new constructor in `TextDocumentIdentifier `makes all LSR tests crash.

So I have implemented the extension method as proposed first.

**Comment by fm-117** [comment]
> Adding a new constructor in `TextDocumentIdentifier `makes all LSR tests crash.
> 
> So I have implemented the extension method as proposed first.

It works but you have to declare the constructor as `internal`.

**Review by fm-117 — CHANGES_REQUESTED** [review]
See my remark on the constructor accessibility

File: TypeCobol.LanguageServer/TypeCobolServer.cs [hunk 2]
```diff
@@ -11,6 +11,7 @@
 using TypeCobol.Compiler.Preprocessor;
 using TypeCobol.LanguageServer.Context;
 using TypeCobol.LanguageServer.SignatureHelper;
+using TypeCobol.LanguageServer.Utilities;
```
Comment (fm-117) [inline TypeCobol.LanguageServer/TypeCobolServer.cs:14]: Not needed

File: TypeCobol.LanguageServer/VsCodeProtocol/Text Document/TextDocumentIdentifier.cs [hunk 3]
```diff
@@ -23,5 +23,17 @@ public TextDocumentIdentifier(string uri)
         {
             this.uri = uri;
         }
+
+        /// <summary>
+        /// Creates a new TextDocumentIdentifier literal.
+        /// @param uri The document's uri.
```
Comment (fm-117) [inline TypeCobol.LanguageServer/VsCodeProtocol/Text Document/TextDocumentIdentifier.cs:29]: Java syntax, we're speaking C# here 💥

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol.LanguageServer/TypeCobolServer.cs (hunk 2 above)
Comment (efr15) [reply TypeCobol.LanguageServer/TypeCobolServer.cs]: Removed

File: TypeCobol.LanguageServer/VsCodeProtocol/Text Document/TextDocumentIdentifier.cs (hunk 3 above)
Comment (efr15) [reply TypeCobol.LanguageServer/VsCodeProtocol/Text Document/TextDocumentIdentifier.cs]: Changed here (but there are many @param in other classes)

**Review by fm-117 — APPROVED** [review]

**Pull request merged**